structure_tests/drivers/docker_driver.go

```diff
@@ -0,0 +1,341 @@
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+
+	f, err := os.Stat(file)
+	if err != nil {
+		return nil, err
+	}
+	if f == nil {
+		return nil, nil
+	}
+	return f, nil
+}
+
+func (d DockerDriver) ReadFile(t *testing.T, path string) ([]byte, error) {
+	file, err := d.retrieveFile(t, path, false)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+	return ioutil.ReadFile(file)
+}
+
+func (d DockerDriver) ReadDir(t *testing.T, path string) ([]os.FileInfo, error) {
+	// TODO(nkubala): unimplemented
+	return nil, nil
+}
+
+// This method takes a command (in the form of a list of args), and does the following:
+// 1) creates a container, based on the "current latest" image, with the command set as
+// the command to run when the container starts
+// 2) starts the container
+// 3) commits the container with its changes to a new image,
+// and sets that image as the new "current image"
+func (d DockerDriver) runAndCommit(t *testing.T, env []string, command []string) string {
+
+	// this is a placeholder command since apparently the client doesnt allow creating
+	// a container without a command.
+	// TODO(nkubala): figure out how to remove this
+	if len(command) == 0 {
+		command = []string{"/bin/sh"}
+	}
+
+	ctx := context.Background()
+
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        currentImage,
+			Env:          env,
+			Cmd:          command,
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+		Context:          ctx,
+	})
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+		return ""
+	}
+
+	err = d.cli.StartContainer(container.ID, nil)
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+	}
+
+	_, err = d.cli.WaitContainer(container.ID)
+
+	image, err := d.cli.CommitContainer(docker.CommitContainerOptions{
+		Container: container.ID,
+		Context:   ctx,
+	})
+
+	if err != nil {
+		t.Errorf("Error committing container: %s", err.Error())
+	}
+
+	currentImage = image.ID
+	return image.ID
+}
+
+func (d DockerDriver) exec(t *testing.T, env []string, command []string) (string, string, int) {
+	ctx := context.Background()
+
+	// first, start container from the current image
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        currentImage,
+			Env:          env,
+			Cmd:          command,
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+		Context:          ctx,
+	})
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+		return "", "", -1
+	}
+
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+
+	err = d.cli.StartContainer(container.ID, nil)
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+	}
+
+	err = d.cli.PauseContainer(container.ID)
```

Comment:
Why are we pausing?

Reply:
if we don't pause, commands in the container start running before we can attach, and we miss output. this is the only way I could find to get the tests to pass reliably.